src/java/org/apache/cassandra/db/ReadCommand.java

```diff
@@ -632,6 +654,84 @@ private void maybeDelayForTesting()
+            return iterator;
+        final long warnThresholdBytes = DatabaseDescriptor.getLocalReadTooLargeWarningThresholdKb() * 1024;
+        final long abortThresholdBytes = DatabaseDescriptor.getLocalReadTooLargeAbortThresholdKb() * 1024;
+        if (warnThresholdBytes == 0 && abortThresholdBytes == 0)
+            return iterator;
+        class QuerySizeTracking extends Transformation<UnfilteredRowIterator>
+        {
+            private DecoratedKey currentKey;
+            private long sizeInBytes = 0;
+
+            @Override
+            public UnfilteredRowIterator applyToPartition(UnfilteredRowIterator iter)
+            {
+                currentKey = iter.partitionKey();
+                sizeInBytes += ObjectSizes.sizeOnHeapOf(currentKey.getKey());
+                return Transformation.apply(iter, this);
+            }
+
+            @Override
+            protected Row applyToStatic(Row row)
+            {
+                return applyToRow(row);
+            }
+
+            @Override
+            protected Row applyToRow(Row row)
+            {
+                addSize(row.unsharedHeapSize());
+                return row;
+            }
+
+            @Override
+            protected RangeTombstoneMarker applyToMarker(RangeTombstoneMarker marker)
+            {
+                addSize(marker.unsharedHeapSize());
+                return marker;
+            }
+
+            @Override
+            protected DeletionTime applyToDeletion(DeletionTime deletionTime)
+            {
+                addSize(deletionTime.unsharedHeapSize());
+                return deletionTime;
+            }
+
+            private void addSize(long size)
+            {
+                this.sizeInBytes += size;
+                if (abortThresholdBytes != 0 && this.sizeInBytes >= abortThresholdBytes)
+                {
+                    String msg = String.format("Query %s attempted to read %d bytes but max allowed is %d; query aborted  (see local_read_too_large_abort_threshold_kb)",
+                                               ReadCommand.this.toCQLString(), this.sizeInBytes, abortThresholdBytes);
+                    Tracing.trace(msg);
+                    MessageParams.remove(ParamType.LOCAL_READ_TOO_LARGE_WARNING);
+                    MessageParams.add(ParamType.LOCAL_READ_TOO_LARGE_ABORT, this.sizeInBytes);
+                    throw new LocalReadSizeTooLargeException(msg);
+                } else if (warnThresholdBytes != 0 && this.sizeInBytes >= warnThresholdBytes)
+                {
+                    MessageParams.add(ParamType.LOCAL_READ_TOO_LARGE_WARNING, this.sizeInBytes);
+                }
+            }
+
+            @Override
+            protected void onClose()
+            {
+                ColumnFamilyStore cfs = Schema.instance.getColumnFamilyStoreInstance(metadata().id);
+                if (cfs != null)
```

Comment:
nit: Is there a case where the `cfs` should be null?

Reply:
two cases

1) unit tests; I think this is why I handled this in the other code path which update metrics
2) race condition with drop table

its mostly to be defensive as null is a valid response =)

Reply:
It seems like a case where a "no-spammed" WARN level message might be useful, but I won't push too hard for it. (ex. Someone is still issuing queries against a table that's being dropped. It's not clearly wrong, but it isn't something you'd want to be doing either...)

Reply:
I think we have issues with dropping tables concurrently with queries, so you will hit an issue anyways.  I wouldn't want to handle here as this is a systemic issue, and these checks are opt-in so can't rely on those log checks